language/move-compiler/src/parser/parsed_tree/ast.rs

```diff
@@ -0,0 +1,625 @@
+pub struct PackageDefinition {
+    pub package: Option<Symbol>,
+    pub named_address_map: NamedAddressMapIndex,
+    pub source_trees: Vec<Definition>,
+    pub source_tokens: Vec<Token>,
+    pub file_hash: FileHash,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SpannedWithComment<T> {
+    pub value: T,
+    pub token_range: TokenRange,
+}
+
+impl<T> SpannedWithComment<T> {
+    pub fn new(value: T, token_range: TokenRange) -> Self {
+        SpannedWithComment { value, token_range }
+    }
+
+    pub fn loc(&self, tokens: &[Token]) -> Loc {
+        self.token_range.loc(tokens)
+    }
+
+    pub fn value(&self) -> &T {
+        &self.value
+    }
+
+    fn is_comment_spaces(token: Tok) -> bool {
+        matches!(
+            token,
+            Tok::Comment(_) | Tok::Space | Tok::Tab | Tok::LF | Tok::CR
+        )
+    }
+
+    fn find_comments(tokens: &[Token]) -> Vec<Comment> {
+        tokens
+            .iter()
+            .filter_map(|t| match t.kind {
+                Tok::Comment(kind) => Some(Comment::new(kind, t.content, t.loc)),
+                _ => None,
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub fn trailing_comments(&self, source_tokens: &[Token]) -> Vec<Comment> {
+        let end_pos = source_tokens[self.token_range.start..self.token_range.end]
+            .iter()
+            .rev()
+            .position(|t| !Self::is_comment_spaces(t.kind));
+        match end_pos {
+            Some(end) => {
+                let trailing_tokens =
+                    &source_tokens[self.token_range.end - end..self.token_range.end];
+                Self::find_comments(trailing_tokens)
+            }
+            None => vec![],
+        }
+    }
+
+    pub fn leading_comments(&self, source_tokens: &[Token]) -> Vec<Comment> {
+        let end_pos = source_tokens[self.token_range.start..self.token_range.end]
+            .iter()
+            .position(|t| !Self::is_comment_spaces(t.kind));
+        match end_pos {
+            Some(end) => {
+                let trailing_tokens =
+                    &source_tokens[self.token_range.start..self.token_range.start + end];
+                Self::find_comments(trailing_tokens)
+            }
+            None => vec![],
+        }
+    }
+}
+
+pub type ParsedToken = SpannedWithComment<Token>;
+
+pub type Name = ParsedToken;
+
+pub type Attribute = SpannedWithComment<Attribute_>;
+pub type Attributes = SpannedWithComment<Vec<Attribute>>;
+pub type Ability = ParsedToken;
+pub type Var = ParsedToken;
+pub type Field = ParsedToken;
+pub type Op = ParsedToken;
+
+// (<Name>|<Num>)::<Name>::<Name>
+pub type NameAccessChain = SpannedWithComment<Vec<Name>>;
+pub type LeadingNameAccess = Name;
+pub type SpecApplyFragment = ParsedToken;
+
+pub type QuantKind = SpannedWithComment<QuantKind_>;
+pub type BinOp = SpannedWithComment<BinOp_>;
+pub type UnaryOp = SpannedWithComment<UnaryOp_>;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ParsedTree {
```

Comment:
This is going to need to be generalized a lot more (as briefly discussed in discord). The goal here is to fully capture the syntax that is in the file, with all whitespace and comment information. It can also be helpful to be able to represent invalid trees (or at least partially invalid trees). As such, parse trees typically do not use distinct types for different syntactic groups

Reply:
But I was having some parsing conflicts with 

```move
spec <Identifier>{
...
}
```
 which could be parsed into `Module` representation or `SpecBlock` representation.
 
And I want to distinguish `Module` and `SpecBlock` because the items inside Module and SpecBlock are different and I want to distinguish them. So I fall back into different syntactic groups.

Hm.... Maybe I should just take `Module` and `SpecBlock` as the same enum and distinguish them later in the translation steps. 

Do you have any preferences on the point above?

Reply:
You will probably have to parse it more generally, and then separate it into distinct AST nodes in this separate translation step I described above. 

In short, expect this parse tree to be _very_ general. It will more or less correspond to the structure of the current parser, but not any of the Rust code inside that structure. All the more complex/contextual analysis will need to be done in this pass that converts the parse tree to the abstract syntax tree

Reply:
@tnowacki Please take a look whether the pr is on the right track or not~!